experimental/gradually-typed/src/Torch/GraduallyTyped/Examples/TwoLayerNetwork.hs

```diff
@@ -0,0 +1,407 @@
+  ) =>
+  HasForward
+    (TwoLayerNetwork fstLayer sndLayer)
+    (Tensor gradient layout dataType device shape)
+    generatorDevice
+    output
+    generatorOutputDevice
+  where
+  forward TwoLayerNetwork {..} input =
+    runIxStateT $
+      ireturn input
+        >>>= IxStateT . forward fstLayer
+        >>>= ireturn . Torch.GraduallyTyped.tanh
+        >>>= IxStateT . forward sndLayer
+
+-- | Options for the Adam optimizer.
+data AdamOptions = AdamOptions
+  { -- | learning rate
+    learningRate :: Double,
+    -- | beta1
+    beta1 :: Double,
+    -- | beta2
+    beta2 :: Double,
+    -- | epsilon
+    epsilon :: Double,
+    -- | weight decay
+    weightDecay :: Double,
+    -- | use amsgrad
+    amsgrad :: Bool
+  }
+
+-- | Default Adam options.
+defaultAdamOptions :: AdamOptions
+defaultAdamOptions =
+  AdamOptions
+    { learningRate = 0.001,
+      beta1 = 0.9,
+      beta2 = 0.999,
+      epsilon = 1e-8,
+      weightDecay = 0.0,
+      amsgrad = False
+    }
+
+-- | Create a new Adam optimizer.
+mkAdam ::
+  -- | Adam options
+  AdamOptions ->
+  -- | initial model parameters
+  [ForeignPtr ATen.Tensor] ->
+  -- | Adam optimizer
+  IO (ForeignPtr ATen.Optimizer)
+mkAdam AdamOptions {..} =
+  ATen.cast7 ATen.adam learningRate beta1 beta2 epsilon weightDecay amsgrad
+
+mseLoss ::
+  forall m gradient layout device dataType shape gradient' layout' device' dataType' shape'.
+  MonadThrow m =>
+  -- | prediction tensor
+  Tensor gradient layout device dataType shape ->
+  -- | target tensor
+  Tensor gradient' layout' device' dataType' shape' ->
+  -- | output tensor
+  m
+    ( Tensor
+        (gradient <|> gradient')
+        (layout <+> layout')
+        (device <+> device')
+        (dataType <+> dataType')
+        (Seq (shape <+> shape') ('Shape '[]))
+    )
+mseLoss prediction target =
+  unsafeThrowableIO $
+    ATen.cast3
+      ATen.mse_loss_ttl
+      prediction
+      target
+      (1 :: Int) -- reduce mean
+
+computeLoss ::
+  _ =>
+  -- | device
+  SDevice device ->
+  -- | model specification
+  ModelSpec model ->
+  -- | model state dict
+  StateDict ->
+  -- | input
+  [[Float]] ->
+  -- | expected output
+  [[Float]] ->
+  -- | random generator
+  Generator generatorDevice ->
+  -- | loss
+  m (_, Generator generatorOutputDevice)
+computeLoss device modelSpec stateDict input expectedOutput =
+  runIxStateT $
+    ilift
+      ( let inputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device input
+            expectedOutputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device expectedOutput
+            model = flip evalStateT stateDict $ fromStateDict modelSpec mempty
+         in (,,) <$> inputTensor <*> expectedOutputTensor <*> model
+      )
+      >>>= ( \(inputTensor, expectedOutputTensor, model) ->
+               ireturn inputTensor
+                 >>>= IxStateT . forward model
+                 >>>= ilift . (`mseLoss` expectedOutputTensor)
+           )
```

Comment:
In this case, what is IxStateT's state? Is it Generator?
Could we write this funciton just like this?
```
computeLoss device modelSpec stateDict input expectedOutput =
let inputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device input
     expectedOutputTensor = sToTensor (SGradient SWithoutGradient) (SLayout SDense) device expectedOutput
     model = flip evalStateT stateDict $ fromStateDict modelSpec mempty
in mseLoss <$> forward model inputTensor <*> pure  expectedOutputTensor
```

Reply:
I pushed the first program that worked. I've since rewritten this part. Yes, this can be simplified.

I'm reaching for the indexed state monad IxStateT by default because the type of the generator state can change. Forward functions can move the generator to a different device.